### PR TITLE
Fix copy_ broadcast behavior on mps

### LINKED ISSUE #105277: Copying to a mps tensor produces incorrect results
### 🐛 Describe the bug

While working on fixing a broken test for `torch.linalg.vander` in #104688, because it should be enabled by cumprod, I noticed that the operator was occasionally inconsistent with the cpu version. I tracked down the inconsistency to a bug in copying to a tensor. After distilling it down here is a small repro:

```
#include <cassert>
#include <torch/torch.h>

torch::Tensor computeResult(torch::DeviceType device) {
    std::cout << "--computing result--\n" << std::endl;
    auto x = torch::arange(2, torch::TensorOptions().device(device));
    std::cout << "x (" << device << ")" << "\n" << x << std::endl;
    auto y = torch::range(0, 3, torch::TensorOptions().device(device)).reshape({2, 2});
    std::cout << "y (" << device << ")" << "\n" << y << std::endl;
    auto x_unsqueeze = x.unsqueeze(1);
    std::cout << "x_unsqueeze (" << device << ")" << "\n" << x_unsqueeze << std::endl;
    y.copy_(x_unsqueeze);
    std::cout << "y after copy" << "\n" << y << std::endl;
    return y;
}

int main() {
    auto cpuResult = computeResult(torch::kCPU);
    auto mpsResult = computeResult(torch::kMPS).to(at::Device(torch::kCPU));
    assert(cpuResult.equal(mpsResult));
    return 0;
}
```

Running the following code I see:
```
--computing result--

x (cpu)
 0
 1
[ CPULongType{2} ]
y (cpu)
 0  1
 2  3
[ CPUFloatType{2,2} ]
x_unsqueeze (cpu)
 0
 1
[ CPULongType{2,1} ]
y after copy
 0  0
 1  1
[ CPUFloatType{2,2} ]
--computing result--

x (mps)
 0
 1
[ MPSLongType{2} ]
y (mps)
 0  1
 2  3
[ MPSFloatType{2,2} ]
x_unsqueeze (mps)
 0
 1
[ MPSLongType{2,1} ]
y after copy
 0  1
 1  3
[ MPSFloatType{2,2} ]
Assertion failed: (cpuResult.equal(mpsResult)), function main, file repro.cpp, line 20.
```

It appears that the copy of the unsqueezed tensor is broken in mps.

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230526
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.24.2
Libc version: N/A

Python version: 3.9.12 (main, Apr  5 2022, 01:52:34)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0.dev20230526
[pip3] torchaudio==2.1.0.dev20230527
[pip3] torchvision==0.16.0.dev20230526
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230526          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230527          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230526          pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by Mr4k**
Fixes #105277

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Copy.mm [hunk 1]
```diff
@@ -20,7 +20,7 @@
   return (void*)alignedAddress;
 }
 
-// Copy sourceBuffer into destBuffer, casting sourceBuffer to src.scalar_type().
+// Copy sourceBuffer into destBuffer, casting sourceBuffer to dst.scalar_type().
```
Comment (Mr4k) [inline aten/src/ATen/native/mps/operations/Copy.mm:23]: drive by comment fix

File: aten/src/ATen/native/mps/operations/Copy.mm [hunk 2]
```diff
@@ -290,8 +290,19 @@ void copy_blit_mps(void* dst, const void* src, size_t size) {
   if (dst.numel() == 0) {
     dst.resize_as_(src);
   }
+
+  TORCH_CHECK(dst.dim() >= src.dim());
   if (dst.dim() > src.dim()) {
     needs_broadcasting = true;
+  } else {
+    const IntArrayRef src_sizes = src.sizes();
```
Comment (Mr4k) [inline aten/src/ATen/native/mps/operations/Copy.mm:298]: copy_ appears to have [broadcasting semantics](https://pytorch.org/docs/stable/notes/broadcasting.html)

File: aten/src/ATen/native/mps/operations/Copy.mm [hunk 3]
```diff
@@ -290,8 +290,19 @@ void copy_blit_mps(void* dst, const void* src, size_t size) {
...
+    const IntArrayRef dst_sizes = dst.sizes();
+    for (const auto j : c10::irange(src.dim())) {
+      if (src_sizes[j] == 1 && dst_sizes[j] > 1) {
```
[the remark this answers is not in the document]
Comment (Mr4k) [reply aten/src/ATen/native/mps/operations/Copy.mm:301]: okay turns out this assumption is not true! I will change this condition to `!=1` and write some additional tests

**Comment by Mr4k** [comment]
@kulinseth bumping this

**Comment by Mr4k** [comment]
@kulinseth @malfet bumping this one more time

File: aten/src/ATen/native/mps/operations/Copy.mm (hunk 1 above)
Comment (Mr4k) [reply aten/src/ATen/native/mps/operations/Copy.mm:23]: after looking at this again, this is incorrect! Sorry I'll remove this

**Review by malfet — APPROVED** [review]
Thank you for the fix

**Pull request closed without merging**